addons/mail_group/models/mail_group.py

```diff
@@ -0,0 +1,319 @@
+        values['alias_force_thread_id'] = self.id
+        values['alias_defaults'] = literal_eval(self.alias_defaults or '{}')
+        return values
+
+    @api.onchange('moderation')
+    def _onchange_moderation(self):
+        if self.moderation and self.env.user not in self.moderator_ids:
+            self.moderator_ids |= self.env.user
+
+    # CONSTRAINTS
+
+    @api.constrains('moderator_ids')
+    def _check_moderator_email(self):
+        if any(not moderator.email for group in self for moderator in group.moderator_ids):
+            raise ValidationError(_('Moderators must have an email address.'))
+
+    @api.constrains('moderation_notify', 'moderation_notify_msg')
+    def _check_moderation_notify(self):
+        if any(group.moderation_notify and not group.moderation_notify_msg for group in self):
+            raise ValidationError(_('The notification message is missing.'))
+
+    @api.constrains('moderation_guidelines', 'moderation_guidelines_msg')
+    def _check_moderation_guidelines(self):
+        if any(group.moderation_guidelines and not group.moderation_guidelines_msg for group in self):
+            raise ValidationError(_('The guidelines description is missing.'))
+
+    @api.constrains('moderator_ids', 'moderation')
+    def _check_moderator_existence(self):
+        if any(not group.moderator_ids for group in self if group.moderation):
+            raise ValidationError(_('Moderated group must have moderators.'))
+
+    @api.model
+    def message_update(self, message_dict=None):
+        """Add the method to make the mail gateway flow work with this model."""
+        return
+
+    def message_post(self, **kwargs):
+        self.ensure_one()
+        # First create the <mail.message>
+        Mailthread = self.env['mail.thread']
+        values = dict((key, val) for key, val in kwargs.items() if key in self.env['mail.message']._fields)
+        author_id, email_from = Mailthread._message_compute_author(kwargs.get('author_id'), kwargs.get('email_from'), raise_exception=True)
+        values.update({
+            'author_id': author_id,
+            'email_from': email_from,
+            'model': self._name,
+            'res_id': self.id,
+            'partner_ids': [],
+        })
+        attachments = kwargs.get('attachments') or []
+        attachment_ids = kwargs.get('attachment_ids') or []
+        attachement_values = Mailthread._message_post_process_attachments(attachments, attachment_ids, values)
+        values.update(attachement_values)
+        mail_message = Mailthread._message_create(values)
+
+        # Create the group message associated
+        group_message = self.env['mail.group.message'].create({
+            'mail_group_id': self.id,
+            'mail_message_id': mail_message.id,
+        })
+
+        # Check the moderation rule to determine if we should accept or reject the email
+        email_normalized = email_normalize(kwargs.get('email_from'))
```

Comment:
Would check email_from is given (and probably raise a ValueError if not), as this is quite mandatory in our flow.